feat: Hybrid unified/recurrent cache

gabe-l-hart opened the pull request:
## Description

This implementation covers both `llama_memory_i` and `llama_kv_cache` interfaces, but they could very well not be correct.

## Discussion

I'm putting this up for discussion even though it doesn't have much value as standalone. My ultimate goal is support for the just-released [granite 4](https://huggingface.co/ibm-granite/granite-4.0-tiny-preview) which is a combination of `mamba2` and `granitemoeshared` layers. I opened https://github.com/ggml-org/llama.cpp/issues/13275 to track the full scope of model architecture changes.

compilade reviewed (commented):
Awesome to see this progress!

gabe-l-hart committed:
feat: Add support for distinguishing recurrent vs non-recurrent layers in hparams

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add c++ side constants for attention layer indices hparam

Branch: GraniteFour

gabe-l-hart committed:
feat: Add llama_model_is_hybrid API call

Also, split llama_model_is_recurrent into llm_arch_is_recurrent in
llama-arch with llama_model_is_recurrent delegating to
llm_arch_is_recurrent. The same split is done for hybird. This is needed
because there are places where the llama_model has not yet been initialized
but we need to check if the model is recurrent (specifically for the
per-layer recurrent check array in hparams).

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Auto-fill hparams.recurrent_layer_arr based on whether the model is recurrent

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented on src/llama-kv-cache.cpp:
```diff
@@ -2384,6 +2384,231 @@ bool llama_kv_cache_recurrent::state_read_data(llama_io_read_i & io, uint32_t ce
+}
+
+llama_ubatch llama_kv_cache_hybrid::ubatch_next(llama_sbatch & sbatch, uint32_t n_ubatch, bool embd_pooled) const {
+    if (m_has_recurrent) {
+        return sbatch.split_simple(n_ubatch);
+    }
```
Ok, I'm following now. I had them backwards in my head

gabe-l-hart commented on src/llama-kv-cache.cpp:
```diff
@@ -2384,6 +2384,231 @@ bool llama_kv_cache_recurrent::state_read_data(llama_io_read_i & io, uint32_t ce
+}
+
+llama_sbatch llama_kv_cache_hybrid::sbatch_init(const llama_batch & batch, bool logits_all) {
+    // If any of the caches are recurrent, require simple split
+    return llama_sbatch(batch, m_hparams.n_embd, m_has_recurrent, logits_all);
```
Thanks for the comment pointer, this is super helpful for understanding what the consequences of these actually are!

gabe-l-hart commented on src/llama-kv-cache.cpp:
```diff
@@ -2384,6 +2384,231 @@ bool llama_kv_cache_recurrent::state_read_data(llama_io_read_i & io, uint32_t ce
+}
+
+bool llama_kv_cache_hybrid::seq_rm(llama_seq_id seq_id, llama_pos p0, llama_pos p1) {
+    // TODO: Will it cause problems if some caches are able to remove the seq
+    //  but others aren't?
```
Ok, interesting. I'm definitely still learning on-the-fly here, but based on this description and the logic [here](https://github.com/ggml-org/llama.cpp/blob/master/tools/server/server.cpp#L3196) in `server.cpp`, it seems like the most correct implementation would be to leak implementation details of the child caches or introduce a new member API for `can_seq_rm` that is `const` but returns the same logic. I think I'll give that a shot and see how far I can get.

gabe-l-hart replied:
Ok, I've pushed an attempt at doing this safely. One thing I noticed is that these mutating methods don't seem to have any sort of locking mechanism, so the way I have it implemented could certainly be prone to thread safety problems if concurrent threads tried to call `seq_rm`. I don't think this is any different than for the current cache implementations since those would also be sensitive to the same races where the validated condition changes after validation but before the members get mutated, but I wanted to double check if this kind of thread safety is guarded against elsewhere (or just assumed to be handled in the client layer).

gabe-l-hart committed:
fix: Use per-layer sizing everywhere in kv caches

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
@compilade I now have a proof-point (#13550) that this works to _some_ extend, though I haven't tested it robustly for edge cases. There are a few additional changes I needed to make on that branch that should _maybe_ come over to this branch, but it gets a little hairy because they interact with adding the actual model architectures. Some possible paths forward:

1. Review this one as-is and take the updates from the Granite 4 branch when they come
2. Try to move the changes to `hparams` to determine per-layer recurrence over to this branch, and then just have unreachable code when initializing the hybrid cache in `llama-model.cpp`
3. Scrap this PR and just review all hybrid cache changes with the incoming models in the Granite 4 branch

Thoughts / Preferences?

gabe-l-hart commented on src/llama-model.h:
```diff
@@ -402,7 +402,10 @@ struct llama_model {
     // note: can mutate `cparams`
     // TODO: move this to new llm_arch_model_i interface
-    llama_memory_i * create_memory(const llama_memory_params & params, llama_cparams & cparams) const;
+    llama_memory_i * create_memory(
```
I just realized that this is entirely unnecessary since `llama_model` already has `hparams` as a member variable 🤦

gabe-l-hart commented:
@compilade @ggerganov I'm going to keep working on some unit tests for the different cache types and try to sort through the CI failure. In the mean time, would it be possible to get an initial review to make sure I'm on the right track?

gabe-l-hart commented:
It looks like the failing tests were all on Windows and therefore fall into [this note](https://github.com/ggml-org/llama.cpp/blob/master/tests/CMakeLists.txt#L105) about not running tests using internal APIs on Windows due to the lack of `LLAMA_API` exports.

ggerganov commented:
I think the hybrid implementation can become similar to the new [iSWA implementation](#13194) - mainly for consistency.

I would like to now simplify the `llama_kv_cache` interface, namely:

- Try to eliminate `set_full()` completely. We can instead `clear()` + init batches with appropriate sizes
- Rework the `sbatch_init()`, `ubatch_next()` and `find_slot()`. It seems that this should be possible to be replaced with a single `init(llama_batch)` call that would prepare all the ubatches beforehand

Will open a draft PR and we can discuss details there.

Adding tests would be really useful - we can merge these from a separate PR without too much delays.

gabe-l-hart committed:
tests: Initial unit tests for memory hierarchy

These only test the basics so far, but should allow for more expansive
tests to come.

Branch: MemoryTests

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
build: Add build step for test-memory on non-windows builds

These tests use private headers, so won't build on windows

Branch: MemoryTests

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
@ggerganov Ported PR with minimal unit tests: https://github.com/ggml-org/llama.cpp/pull/13669

I didn't try to expand them to incorporate the iSWA cache yet, but will look to do that as I go through the changes there and refactor this hybrid cache branch.

gabe-l-hart committed:
feat: Move layer_filter_cb up to llama_kv_cache

This will be needed by other cache types as well, so centralizing the
definition will make it more reusable.

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add layer filter to recurrent cache

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Initial implementation of llama_kv_cache_hybrid

Condensed from initial version
https://github.com/gabe-l-hart/llama.cpp/tree/ec08571

The only difference is the removal of m_layer_cache_map which was unused
and unnecessary now that child caches are instantiated with their own
filters.

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Instantiate hybrid cache for hybrid models (currently none)

This includes a slight architectural change where create_memory now only
uses model architectures in the switch statement if their required cache
type is not handled by llm_arch_is_[recurrent|hybrid].

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
refactor: rename *_is_hybrid -> *_is_hybrid_recurrent

The implementation of the hybrid cache intentionally does not specify the
types of the child caches, so there was a naming mismatch with these
predicate functions that used "hybrid" to imply "hybrid recurrent."

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
@ggerganov This branch is now rebuilt based on the `iSWA` cache. I kept most of the implementation of `llama_kv_cache_hybrid` the way it was where the sub-types of the child caches are not specified to _hopefully_ make it easier to extend this to other per-layer hybrid styles in the future. Like before, I've plumbed through the instantiation of the hybrid cache to the point where any model that answers `true` to `llm_arch_is_hybrid_recurrent` will instantiate a hybrid cache in `create_memory`, but currently there are no models that fit that, so it will not be exercised until the next PR with the concrete hybrid architectures.

gabe-l-hart commented on src/llama-kv-cache.cpp:
```diff
@@ -1898,6 +1916,11 @@ llama_kv_cache_recurrent::llama_kv_cache_recurrent(
     v_l.reserve(n_layer);
 
     for (int i = 0; i < n_layer; i++) {
+        if (filter && !filter(i)) {
```
This is buggy because `push_back` is used below, so later when we index directly into the given per-layer tensor vectors, the final layers will be out-of-bounds reads.

gabe-l-hart committed:
fix: Fix indexing into k_l for recurrent cache with filter

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
I've updated the Granite 4 / Bamba branch (#13550) to test this out and found a couple of bugs that should now be fixed and functional.

gabe-l-hart committed:
fix: Remove unused kv cache methods after rebase

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix(tests): Fix constructors in tests for signature changes after rebase

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
tests(wip): More robust test for unified cache

I'm still not clear how cache hits should be detected since find_slot does
not seem to take into account the tokens themselves and simply looks for a
sequence of cells that fits the size of the ubatch and has no set positions
in any of the cells. I'm clearly still missing something about how this
works!

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix(tests): Fix constructors in tests for signature changes after rebase

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add can_seq_rm API to llama_kv_cache API

This will be key for the hybrid cache which needs to be able to validate
that all children can perform seq_rm cleanly before attempting to remove
the seq from any single child to avoid ending up in a corrupted state.

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add split_equal to init(...) signature

This will enable the hybrid cache to control the split type for all
children together.

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Overhaul hybrid cache for refactor part3 (::init interface)

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
tests(wip): Comment out broken test for now and fix other constructor signatures

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>
